File: cli/internal/error/sentry.go
package error

import (
	"log/slog"
	"os"
	"time"

	"github.com/getsentry/sentry-go"
	"github.com/go-errors/errors"
	"github.com/pushtoprodai/prod-cli/internal/config"
)

// client provides opt-in error tracking. It is enabled only when the user sets
// PROD_SENTRY_DSN (their own Sentry); otherwise every method is a no-op.
type client struct {
	initialized bool
}

// new creates a new error tracking client
func newClient() *client {
	return &client{
		initialized: false,
	}
}

// initialize sets up Sentry with the provided DSN
// This should be called early in the application lifecycle
func (c *client) initialize() error {
	if c.initialized {
		return nil // Already initialized
	}

	// Opt-in only: prod never reports to the maintainer. Error tracking is enabled
	// solely when the user points it at THEIR OWN Sentry via PROD_SENTRY_DSN.
	dsn := os.Getenv("PROD_SENTRY_DSN")
	if dsn == "" {
		return nil // telemetry disabled — the default for the local-first binary
	}

	err := sentry.Init(sentry.ClientOptions{
		Dsn: dsn,
		// Set sample rate for performance monitoring
		TracesSampleRate: 0.1,
		// Set the environment (can be overridden by caller if needed)
		Environment: config.GetEnvironment(),
		// Enable debug mode in development
		Debug: false,
		// Disable sending personally identifiable information by default
		SendDefaultPII: false,
		// Set release information if available
		// Release: version, // Can be set by caller
		// Remove server_name and geographic data for privacy since this runs on user machines
		BeforeSend: func(event *sentry.Event, hint *sentry.EventHint) *sentry.Event {
			event.ServerName = ""
			// Remove user IP address and any geographic information
			if event.User.IPAddress != "" {
				event.User.IPAddress = ""
			}
			return event
		},
	})
	if err != nil {
		return errors.WrapPrefix(err, "failed to initialize Sentry", 0)
	}

	c.initialized = true
	return nil
}

// captureError sends an error to the user's own Sentry. It's a no-op unless the
// user opted in via PROD_SENTRY_DSN (c.initialized). Tracking-process errors are
// logged, not returned, so they never disrupt the application flow.
func (c *client) captureError(err error) {
	if !c.initialized {
		slog.Debug("Error tracking: Sentry  not initialized")
		return
	}

	// Capture the error with Sentry
	sentry.CaptureException(err)
}

// captureErrorWithContext sends an error to Sentry with additional context
func (c *client) captureErrorWithContext(err error, context map[string]any) {
	if !c.initialized {
		slog.Debug("Error tracking: Sentry  not initialized")
		return
	}

	// Configure scope with additional context
	sentry.WithScope(func(scope *sentry.Scope) {
		// Add context data
		for key, value := range context {
			// Convert interface{} to map[string]interface{} for Sentry context
			if contextMap, ok := value.(map[string]any); ok {
				scope.SetContext(key, contextMap)
			} else {
				// If not a map, create a simple context with the value
				scope.SetContext(key, map[string]any{"value": value})
			}
		}

		// Capture the error
		sentry.CaptureException(err)
	})
}

// captureMessage sends a message to Sentry (for non-error events)
func (c *client) captureMessage(message string, level sentry.Level) {
	if !c.initialized {
		slog.Debug("Error tracking: Sentry  not initialized")
		return
	}

	// Set level in scope and capture message
	sentry.WithScope(func(scope *sentry.Scope) {
		scope.SetLevel(level)
		sentry.CaptureMessage(message)
	})
}

// flush waits for all events to be sent to Sentry
// Should be called before application exits
func (c *client) flush() {
	if !c.initialized {
		return
	}

	// Wait up to 5 seconds for events to be sent
	sentry.Flush(5 * time.Second)
}

// addBreadcrumb adds a breadcrumb to help with debugging
func (c *client) addBreadcrumb(message, category string, level sentry.Level) {
	if !c.initialized {
		slog.Debug("Error tracking: Sentry  not initialized")
		return
	}

	sentry.AddBreadcrumb(&sentry.Breadcrumb{
		Message:   message,
		Category:  category,
		Level:     level,
		Timestamp: time.Now(),
	})
}

File: supabase/functions/_shared/sentry.ts
// @ts-ignore - Deno Sentry module types
import * as Sentry from 'https://deno.land/x/sentry/index.mjs'

// Deno global declaration for Edge Functions
declare const Deno: {
  env: {
    get(key: string): string | undefined
  }
}

let initialized = false

/**
 * Initialize Sentry for Supabase Edge Functions
 */
export function initSentry(): void {
  if (initialized) {
    return
  }

  const dsn = Deno.env.get('SB_SENTRY_DSN')
  if (!dsn) {
    console.warn('SENTRY_DSN not configured, skipping Sentry initialization')
    return
  }

  // @ts-ignore - Sentry types
  Sentry.init({
    dsn,
    // Disable default integrations to avoid scope sharing between requests
    defaultIntegrations: false,
    // Performance Monitoring
    tracesSampleRate: 0.1,
    // Set sampling rate for profiling - this is relative to tracesSampleRate
    profilesSampleRate: 0.1,
    // Disable sending personally identifiable information by default
    sendDefaultPii: false,
    // Remove server_name and geographic data for privacy
    beforeSend: (event: any) => {
      event.server_name = ''
      // Remove user IP address and any geographic information
      if (event.user && event.user.ip_address) {
        event.user.ip_address = ''
      }
      return event
    },
  })

  // Set region and execution_id as custom tags
  const region = Deno.env.get('SB_REGION')
  const executionId = Deno.env.get('SB_EXECUTION_ID')
  
  if (region) {
    // @ts-ignore - Sentry types
    Sentry.setTag('region', region)
  }
  if (executionId) {
    // @ts-ignore - Sentry types
    Sentry.setTag('execution_id', executionId)
  }

  initialized = true
}

/**
 * Capture an exception with Sentry using isolated scope
 * This prevents context sharing between requests
 */
export function captureException(error: Error, context?: Record<string, any>): void {
  if (!initialized) {
    console.error('Sentry not initialized, logging error:', error)
    return
  }

  // @ts-ignore - Sentry types
  Sentry.withScope((scope: any) => {
    // Add context data if provided
    if (context) {
      for (const [key, value] of Object.entries(context)) {
        scope.setContext(key, typeof value === 'object' ? value : { value })
      }
    }

    // Add function-specific tags
    scope.setTag('function_name', Deno.env.get('SUPABASE_FUNCTION_NAME') || 'unknown')
    scope.setTag('environment', Deno.env.get('ENVIRONMENT') || 'production')

    // @ts-ignore - Sentry types
    Sentry.captureException(error)
  })
}

/**
 * Capture a message with Sentry using isolated scope
 */
export function captureMessage(
  message: string, 
  level: 'fatal' | 'error' | 'warning' | 'info' | 'debug' = 'info', 
  context?: Record<string, any>
): void {
  if (!initialized) {
    console.log('Sentry not initialized, logging message:', message)
    return
  }

  // @ts-ignore - Sentry types
  Sentry.withScope((scope: any) => {
    // Add context data if provided
    if (context) {
      for (const [key, value] of Object.entries(context)) {
        scope.setContext(key, typeof value === 'object' ? value : { value })
      }
    }

    // Add function-specific tags
    scope.setTag('function_name', Deno.env.get('SUPABASE_FUNCTION_NAME') || 'unknown')
    scope.setTag('environment', Deno.env.get('ENVIRONMENT') || 'production')

    scope.setLevel(level)
    // @ts-ignore - Sentry types
    Sentry.captureMessage(message)
  })
}

/**
 * Flush pending Sentry events
 * Should be called before function completes
 */
export async function flushSentry(timeout = 2000): Promise<void> {
  if (!initialized) {
    return
  }

  // @ts-ignore - Sentry types
  await Sentry.flush(timeout)
}

/**
 * Add a breadcrumb for debugging
 */
export function addBreadcrumb(
  message: string, 
  category = 'custom', 
  level: 'fatal' | 'error' | 'warning' | 'info' | 'debug' = 'info'
): void {
  if (!initialized) {
    return
  }

  // @ts-ignore - Sentry types
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    timestamp: Date.now() / 1000,
  })
}
